File: src/code/astar.java
import java.util.*;

class Node {
    String id;
    double x, y;
    double heuristic;
    
    public Node(String id, double x, double y, double heuristic) {
        this.id = id;
        this.x = x;
        this.y = y;
        this.heuristic = heuristic;
    }
}

class Edge {
    String from, to;
    double weight;
    
    public Edge(String from, String to, double weight) {
        this.from = from;
        this.to = to;
        this.weight = weight;
    }
}

class AStarNode implements Comparable<AStarNode> {
    String id;
    double g, h, f;
    String parent;
    
    public AStarNode(String id, double g, double h, String parent) {
        this.id = id;
        this.g = g;
        this.h = h;
        this.f = g + h;
        this.parent = parent;
    }
    
    @Override
    public int compareTo(AStarNode other) {
        return Double.compare(this.f, other.f);
    }
}

class Graph {
    private Map<String, Node> nodes;
    private List<Edge> edges;
    private Map<String, List<Edge>> adjacencyList;
    
    public Graph() {
        nodes = new HashMap<>();
        edges = new ArrayList<>();
        adjacencyList = new HashMap<>();
    }
    
    public void addNode(Node node) {
        nodes.put(node.id, node);
        adjacencyList.putIfAbsent(node.id, new ArrayList<>());
    }
    
    public void addEdge(String from, String to, double weight) {
        Edge edge = new Edge(from, to, weight);
        edges.add(edge);
        adjacencyList.get(from).add(edge);
    }
    
    public Node getNode(String id) {
        return nodes.get(id);
    }
    
    public List<Edge> getNeighbors(String id) {
        return adjacencyList.getOrDefault(id, new ArrayList<>());
    }
}

class AStar {
    private Graph graph;
    private String startId, goalId;
    
    public AStar(Graph graph, String startId, String goalId) {
        this.graph = graph;
        this.startId = startId;
        this.goalId = goalId;
    }
    
    public List<String> findPath() {
        PriorityQueue<AStarNode> openList = new PriorityQueue<>();
        Set<String> closedList = new HashSet<>();
        Map<String, Double> gScore = new HashMap<>();
        Map<String, String> parent = new HashMap<>();
        
        // Initialize start node
        Node startNode = graph.getNode(startId);
        gScore.put(startId, 0.0);
        openList.add(new AStarNode(startId, 0, startNode.heuristic, null));
        parent.put(startId, null);
        
        while (!openList.isEmpty()) {
            AStarNode current = openList.poll();
            
            // Goal reached
            if (current.id.equals(goalId)) {
                return reconstructPath(parent, goalId);
            }
            
            // Skip if already processed
            if (closedList.contains(current.id)) continue;
            closedList.add(current.id);
            
            // Explore neighbors
            for (Edge edge : graph.getNeighbors(current.id)) {
                String neighborId = edge.to;
                
                if (closedList.contains(neighborId)) continue;
                
                double tentativeG = gScore.get(current.id) + edge.weight;
                
                if (!gScore.containsKey(neighborId) || tentativeG < gScore.get(neighborId)) {
                    gScore.put(neighborId, tentativeG);
                    parent.put(neighborId, current.id);
                    
                    Node neighborNode = graph.getNode(neighborId);
                    double h = neighborNode.heuristic;
                    openList.add(new AStarNode(neighborId, tentativeG, h, current.id));
                }
            }
        }
        
        return new ArrayList<>(); // No path found
    }
    
    private List<String> reconstructPath(Map<String, String> parent, String current) {
        List<String> path = new ArrayList<>();
        while (current != null) {
            path.add(0, current);
            current = parent.get(current);
        }
        return path;
    }
}

public class AStarAlgorithm {
    public static void main(String[] args) {
        // Create graph
        Graph graph = new Graph();
        
        // Add nodes with heuristic values
        graph.addNode(new Node("A", 0, 0, 6));
        graph.addNode(new Node("B", 1, 0, 5));
        graph.addNode(new Node("C", 0, 1, 5));
        graph.addNode(new Node("D", 2, 0, 3));
        graph.addNode(new Node("E", 2, 1, 3));
        graph.addNode(new Node("F", 3, 0, 1));
        graph.addNode(new Node("G", 4, 0, 0));
        
        // Add directed edges
        graph.addEdge("A", "B", 2);
        graph.addEdge("A", "C", 5);
        graph.addEdge("B", "D", 3);
        graph.addEdge("C", "E", 2);
        graph.addEdge("D", "F", 4);
        graph.addEdge("E", "F", 1);
        graph.addEdge("D", "E", 2);
        graph.addEdge("F", "G", 3);
        
        // Run A* algorithm
        AStar astar = new AStar(graph, "A", "G");
        List<String> path = astar.findPath();
        
        // Print result
        System.out.print("Path found: ");
        for (int i = 0; i < path.size(); i++) {
            System.out.print(path.get(i));
            if (i < path.size() - 1) System.out.print(" -> ");
        }
        System.out.println();
    }
}

File: src/code/astar.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <cmath>
#include <algorithm>

using namespace std;

struct Node {
    string id;
    double x, y;
    double heuristic;
    
    Node(string id, double x, double y, double h) 
        : id(id), x(x), y(y), heuristic(h) {}
};

struct Edge {
    string from, to;
    double weight;
    
    Edge(string f, string t, double w) 
        : from(f), to(t), weight(w) {}
};

struct AStarNode {
    string id;
    double g, h, f;
    string parent;
    
    AStarNode(string id, double g, double h, string parent = "")
        : id(id), g(g), h(h), f(g + h), parent(parent) {}
    
    bool operator>(const AStarNode& other) const {
        return f > other.f;
    }
};

class Graph {
private:
    unordered_map<string, Node> nodes;
    vector<Edge> edges;
    unordered_map<string, vector<pair<string, double>>> adjacencyList;
    
public:
    void addNode(const Node& node) {
        nodes[node.id] = node;
    }
    
    void addEdge(const string& from, const string& to, double weight) {
        edges.push_back(Edge(from, to, weight));
        adjacencyList[from].push_back({to, weight});
    }
    
    Node getNode(const string& id) {
        return nodes[id];
    }
    
    vector<pair<string, double>> getNeighbors(const string& id) {
        return adjacencyList[id];
    }
};

class AStar {
private:
    Graph& graph;
    string startId, goalId;
    
public:
    AStar(Graph& g, string start, string goal) 
        : graph(g), startId(start), goalId(goal) {}
    
    vector<string> findPath() {
        priority_queue<AStarNode, vector<AStarNode>, greater<AStarNode>> openList;
        unordered_set<string> closedList;
        unordered_map<string, double> gScore;
        unordered_map<string, string> parent;
        
        // Initialize start node
        Node startNode = graph.getNode(startId);
        gScore[startId] = 0;
        openList.push(AStarNode(startId, 0, startNode.heuristic));
        parent[startId] = "";
        
        while (!openList.empty()) {
            AStarNode current = openList.top();
            openList.pop();
            
            // Goal reached
            if (current.id == goalId) {
                return reconstructPath(parent, goalId);
            }
            
            // Skip if already processed
            if (closedList.count(current.id)) continue;
            closedList.insert(current.id);
            
            // Explore neighbors
            for (auto& neighbor : graph.getNeighbors(current.id)) {
                string neighborId = neighbor.first;
                double edgeWeight = neighbor.second;
                
                if (closedList.count(neighborId)) continue;
                
                double tentativeG = gScore[current.id] + edgeWeight;
                
                if (!gScore.count(neighborId) || tentativeG < gScore[neighborId]) {
                    gScore[neighborId] = tentativeG;
                    parent[neighborId] = current.id;
                    
                    Node neighborNode = graph.getNode(neighborId);
                    double h = neighborNode.heuristic;
                    openList.push(AStarNode(neighborId, tentativeG, h, current.id));
                }
            }
        }
        
        return {}; // No path found
    }
    
private:
    vector<string> reconstructPath(unordered_map<string, string>& parent, string current) {
        vector<string> path;
        while (!current.empty()) {
            path.push_back(current);
            current = parent[current];
        }
        reverse(path.begin(), path.end());
        return path;
    }
};

int main() {
    // Create graph
    Graph graph;
    
    // Add nodes with heuristic values
    graph.addNode(Node("A", 0, 0, 6));
    graph.addNode(Node("B", 1, 0, 5));
    graph.addNode(Node("C", 0, 1, 5));
    graph.addNode(Node("D", 2, 0, 3));
    graph.addNode(Node("E", 2, 1, 3));
    graph.addNode(Node("F", 3, 0, 1));
    graph.addNode(Node("G", 4, 0, 0));
    
    // Add directed edges
    graph.addEdge("A", "B", 2);
    graph.addEdge("A", "C", 5);
    graph.addEdge("B", "D", 3);
    graph.addEdge("C", "E", 2);
    graph.addEdge("D", "F", 4);
    graph.addEdge("E", "F", 1);
    graph.addEdge("D", "E", 2);
    graph.addEdge("F", "G", 3);
    
    // Run A* algorithm
    AStar astar(graph, "A", "G");
    vector<string> path = astar.findPath();
    
    // Print result
    cout << "Path found: ";
    for (size_t i = 0; i < path.size(); i++) {
        cout << path[i];
        if (i < path.size() - 1) cout << " -> ";
    }
    cout << endl;
    
    return 0;
}
